### DISCUSSION TITLE
How can a user resize a node (if is tagged as 'resizable')?

### DISCUSSION THREAD

**Discussion by ulissepress** [bcakmakoglu/vue-flow#45] [Q&A] [answered]
The use case flow:

1. The user select a node
2. The user can resize the node using the mouse (using for example the bottom right corner, 
     that will display some kind of handle)

How this can be implemented ?

**Comment by bcakmakoglu** [comment, marked as answer]
To implement a "resizable" node you would have to create a custom node that can support the resizing on a corner click.
Or you could check the onClick event for nodes and calculate where the click happened etc. but that seems more complicated than adding a custom node 😅 

But generally you can change a node's size by applying styles to it, so an example implementation could look like this.
```vue
<script setup>
import { VueFlow } from '@braks/vue-flow'
import CustomNode from './CustomNode.vue'

const elements = ref([
  { id: '1', type: 'input', label: 'Node 1', position: { x: 250, y: 5 } },
  {
    id: '2',
	type: 'custom',
    label: 'Node 2',
    position: { x: 100, y: 100 },
    style: { backgroundColor: 'rgba(255, 0, 255, 0.7)', height: '150px', width: '270px' },
  },
])

const onResize = (size) => {
	const node = elements.value.find((el) => el.id === '2')
	if (node) node.style = { ...node.style, ...size }
}
</script>
<template>
	<VueFlow v-model="elements">
		<template #node-custom="props">
			<CustomNode v-bind="props" @resize="onResize" />
		</template>
	</VueFlow>
</template>
```

```vue
<script setup>
// CustomNode

const emit = defineEmits(['resize'])
const onResize = () => {
	emit('resize', size)
}
</script>
<template>
	<div>...</div>
</template>
```

In 0.4.0 you could actually manipulate the dimensions of a node directly instead of applying styles but the above example would work in 0.3.x or 0.4.0-x.

**Reply by ulissepress** [reply]
Thank your for your prompt reply! 

It means that in CustomNode.vue i have to implement myself the resize of the the 'div' using the mouse etc. and then because this component emits a 'resize' event, this is also reflected in the vue-flow node.

Do you can i use an existing package like [vue-resizable](https://github.com/nikitasnv/vue-resizable) to handle the div resizing ?

**Reply by bcakmakoglu** [reply]
Depending on the complexity of your implementation something like this would also be enough

```vue
<script lang="ts" setup>
import { Position, Handle, ValidConnectionFunc } from '@braks/vue-flow'

interface Props {
  id: string
  connectable?: boolean
  label?: string
  isValidSourcePos?: ValidConnectionFunc
  isValidTargetPos?: ValidConnectionFunc
}

const props = withDefaults(defineProps<Props>(), {
  connectable: false,
  sourcePosition: 'bottom' as Position,
  targetPosition: 'top' as Position,
})
</script>
<script lang="ts">
export default {
  name: 'ResizableNode',
  inheritAttrs: false,
}
</script>
<template>
  <Handle
    type="target"
    :position="props.targetPosition"
    :is-connectable="props.connectable"
    :is-valid-connection="props.isValidTargetPos"
  />
  <div class="resize-node">
    <div v-html="props.label"></div>
    <div ref="el" class="resizer nodrag" />
  </div>
  <Handle
    type="source"
    :position="props.sourcePosition"
    :is-connectable="props.connectable"
    :is-valid-connection="props.isValidSourcePos"
  />
</template>
<style>
.resize-node {
  padding: 1rem;
  border: 1px solid gray;
  border-radius: 4px;
  background: white;
}

.resizer {
  resize: both;
  width: 200px;
  height: 200px;
  outline: none;
  white-space: pre;
  overflow-wrap: normal;
  overflow: hidden;
  background: aliceblue;
  border: 1px solid gray;
  border-radius: 4px;
}
</style>
```

This would create a node that is just resizable by using CSS, no extra JS necessary.

You can of course use something like `vue-resizable` or try [vueuse](https://vueuse.org/core/useelementsize/)

**Reply by ulissepress** [reply]
Great !

May make sense for you to add in the next release a new attribute, at node level, **resizable: "no" | "horizontal" | "vertical" | "both"**, so it becomes a 'standard' feature of Vue Flow ? 

In fact i am submitting a [new issue for feature request](https://github.com/bcakmakoglu/vue-flow/issues/47) :-)

**Reply by ghiscoding** [reply]
@bcakmakoglu Thanks for providing this solution, I do have a few small issues related to this approach though. Every node would require this extra <div> wrapper around the node and only that extra div is draggable & selectable which mean that clicking from inside the node won't be enough to drag the node (we need to click on that resizer padding). This brings a second problem, prior to this change, we could click anywhere on the node to drag it and that would add a "selected" class to the parent node, however this no longer happen when clicking inside the node. I have styling applied to that "selected" class and I wish to keep it, so I came up with this code (which works but it's not pretty)
```ts
// remove "selected" class
document.body.onclick = () => document.querySelector('.resize-node')?.parentElement?.classList.remove('selected');

// apply "selected" class
function cardClicked(e: Event) {
    e.stopImmediatePropagation(); // stop bubbling up to body
    document.querySelector('.resize-node')?.parentElement?.classList.add('selected');
}
```
```html
<template>
    <div class="resize-node">
        <q-card class="resizable nodrag" @click="cardClicked">
```
but it's not exactly great to have that much code to add back the "selected" class. Are there better ways to deal with this issue? I also wonder if I could somehow do the same to make the node draggable again from anywhere? I wish to have simpler code that is easy to use for all nodes and doesn't lose features that I had before the resizer code. I guess all these issues are related to the fact that we need to add the CSS class `"nodrag"` to the resizer portion but it'd be nice to have a solution that doesn't remove all of that no drag related feature.

**Reply by bcakmakoglu** [reply]
I haven't looked into this node in a while tbh so I'm not 100% sure how it worked but you should be good by adding a `dragHandle` element to your node definition (a query selector works, like `.custom-node` for classes or `#custom-node` for ids).

Setting selected on a node can be done by just toggling it's selected property. 
```js
const node = findNode(someId)

node.selected = true
```

**Reply by ghiscoding** [reply]
Thanks, the `findNode` is cleaner than using straight up query selector, but I'm not exactly sure what you mean about the `dragHandle`, now the updated code I have is this

```ts
document.body.onclick = () => changeNodeSelection(false);

function cardClicked(e: Event) {
    e.stopImmediatePropagation();
    changeNodeSelection(true);
}

function changeNodeSelection(selected: boolean) {
    const node = findNode(props.id);
    if (node) {
        node.selected = selected;
    }
}
```

**Reply by bcakmakoglu** [reply]
You can specify a drag handle for your nodes, which will enable the drag handler when the mouse is over that element.

```js
const nodes = [
	{
		id: '1',
		label: 'Node 1',
		position: { x: 0, y: 0 },
		type: 'custom',
		dragHandle: '.node-drag-handle'
	}
]
```

```vue
<template>
	<div>
		<Handle .. />
		<div class="node-drag-handle">
			Drag me
		</div>
		<Handle ... />
	</div>
</template>

**Reply by ghiscoding** [reply]
hmm ok but unless I get the naming confused again, isn't a Handle just where we connect the lines, how is that going to help with Custom Node dragging?

**Reply by ghiscoding** [reply]
I see what you mean now, however the custom handle wasn't necessary. I just added a div on top of the node with that text and styled it a bit. I also use the div wrapper with ".resize-node" css class but made it's border and background color as transparent, so user can drag the node from anywhere around the node (the padding area between the node and its wrapper).

```html
<div class="drag-handle">Drag me <q-icon name="drag_indicator" /></div>
```
```scss
.drag-handle {
    background-color: lightgray;
    border-radius: 4px 4px 0 0;
    width: fit-content;
    padding: 0 5px;
    position: absolute;
    right: 12px;
    top: -12px;
}
```

![image](https://user-images.githubusercontent.com/643976/202267751-c4852f05-70d4-44c8-92aa-2ba327b43d36.png)


it's not too bad but still wish to find a better alternative that would allow me to drag the node from anywhere and still keep the resize. I'll probably stick with that for now until someone else comes with a better solution. Thanks

**Reply by bcakmakoglu** [reply]
The naming could be confusing but a drag handle and the `Handle` component are two different things.
A *drag* handle is just used for dragging - by default the whole node is basically the drag handle. But it can be any element in your node.
It's completely unrelated to `Handle` which is related to connecting nodes, not dragging ^^

Currently I am working on something else but I should have some free time soon and I'll go look into doing some improvements to the Resize-Rotate-Node. Just can't promise when it'll be done and to what extend I can improve it 😅
